frontend/src/pages/CompareV2.tsx

```diff
@@ -158,6 +196,90 @@ function getRunArtifacts(runs: ApiRunDetail[], mlmdPackages: MlmdPackage[]): Run
   });
 }
 
+interface VisualizationPlaceholderProps {
+  metricsTabText: string;
+}
+
+function VisualizationPlaceholder(props: VisualizationPlaceholderProps) {
+  const { metricsTabText } = props;
+  return (
+    <div className={classes(css.visualizationPlaceholder)}>
+      <p className={classes(css.visualizationPlaceholderText)}>
+        The selected {metricsTabText} will be displayed here.
+      </p>
+    </div>
+  );
+}
+
+interface MetricsDropdownProps {
+  filteredRunArtifacts: RunArtifacts[];
+  metricsTabText: string;
+}
+
+function MetricsDropdown(props: MetricsDropdownProps) {
+  const { filteredRunArtifacts, metricsTabText } = props;
+  const [selectedItem, setSelectedItem] = useState<SelectedItem>({ itemName: '', subItemName: '' });
+
+  const dropdownItems: DropdownItem[] = [];
+  for (const x of filteredRunArtifacts) {
+    const subItems: DropdownSubItem[] = [];
+    for (const y of x.executionArtifacts) {
+      const executionName = y.execution
+        .getCustomPropertiesMap()
+        .get('display_name')
+        ?.getStringValue();
+      if (executionName) {
+        const executionLinkedArtifacts: DropdownSubItem[] = [];
+        for (const z of y.linkedArtifacts) {
+          const artifactName = z.event
+            .getPath()
+            ?.getStepsList()[0]
+            .getKey();
```

Comment:
It's also possible to get the artifact name through `artifact.getCustomPropertiesMap()`:

https://github.com/kubeflow/pipelines/blob/c5df3e88cc48580e7c4ffba03a163e9fddbaf3fd/frontend/src/components/tabs/RuntimeNodeDetailsV2.tsx#L239-L243

Is one method preferable to the other? (If I retrieve the name from the `artifact`, I might not need to store the `event` information.)

Additionally, while I log a warning when I cannot find the `artifactName` here, the referenced code simply replaced the name with `-`. Is one solution preferable to the other? (In what circumstances would the name not be found?)

Reply:
1. The artifact name in `RuntimeNodeDetailsV2` probably can be updated using the name in `Event` as well. Can you validate if the name in `event` is the same as name in `artifact`? If not, then we use the name in `Event` as source of truth.
2. I think the combination of logging a warning and replace with `-` is preferred. Having a `-` allows people to click on it in case it represents a hyperlink.

Reply:
1. When comparing the metrics in my local examples, they all matched. I found the following comment: https://github.com/kubeflow/pipelines/blob/9d3ac2692dcbf6272ad6c83afa5977b871655404/frontend/src/mlmd/MlmdUtils.ts#L400-L401
Is there a difference between the `display_name` and the "actual name"? If so, since I am displaying the artifact name to the frontend, would I want to use the `display_name`?
Some files like `MetricsVisualizations.tsx` use the `display_name` method when displaying the name, while other files use the following helper function: https://github.com/kubeflow/pipelines/blob/9d3ac2692dcbf6272ad6c83afa5977b871655404/frontend/src/mlmd/MlmdUtils.ts#L380-L389.
I think I may want to use the `display_name` method, but in the meantime I replaced my usage of the function with the helper function I found above to simplify the code: https://github.com/kubeflow/pipelines/pull/7943/commits/55eb34de5f6348b6f9454512e07a81086f0cd715.

2. The only complication with this method is that the `TwoLevelDropdown` implementation currently assumes that each item name is unique, so it will not be able to identify items correctly when they are selected if `-` is used. I could add unique IDs for each of these items, but users will still not be able to identify them by name.
What is this case with a hyperlink? If the artifact represents a hyperlink, would the visualization information (for confusion matrices, HTML, and Markdown) still be available?

Reply:
1. Sounds good on using getArtifactName() from MlmdUtils. The reason is that DAG visualization combines both IR(PipelineSpec) and Runtime information(MLMD), so I am using artifact `display_name` in some cases. But ideally we should use `Event` whenever possible.
2. That is good point, I think `using unique ID when name is not available` sounds a good idea in this PR. The hyperlink scenario assumes that you have already clicked on an Artifact node on DAG, so there is not much room for ambiguity. Hyperlink redirects user to the `Artifact detail page` and there is no metrics visualization in this detail page.

Reply:
1. Ok, sounds great!
2. I want to clarify this since the implementation will affect future PRs as well - if the name is not available, is there still a chance I can display a visualization with that artifact? If not, I think I want to filter those artifacts out of the dropdown because they will be useless when the user attempts to view the relevant visualization. If yes, I will include them and use `run.run!.id!`, `execution.getId()`, and `artifact.getId()` to allow the user to uniquely identify each option.

Reply:
2. Yes, even if the name is not available, it is still possible that you can display a visualization with that artifact. So you make a good point to use Ids to identify option, but we can simplify it by using `artifactId` instead of `runId + executionId + artifactId`, because artifactId is already unique across the whole MLMD store. And it will make the dropdown easier to show options as well, because you can just append artifact ID after artifact name. Even if the artifact name can be `-`, it can still use `artifactId` to distinguish each option.

Reply:
Ok, I resolved this issue based on our offline discussion. We decided on the following logic:
1. If the `run name` is not available, log a warning and do not show that dropdown item.
2. If the `execution name` is not available, log a warning and replace it with the `execution ID`.
3. If the `artifact name` is not available, log a warning and replace it with the `artifact ID`.

This allows us to keep the same `TwoLevelDropdown` implementation, and the text within each dropdown item is still distinct, allowing users to identify the relevant visualization.

I also split up the function further to improve readability, and added some checks to an existing test to verify functionality.

The commit is here: https://github.com/kubeflow/pipelines/pull/7943/commits/c40e00f81b3d5ea78aac29afa03c9aa3ea553570.